crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,529 @@
+//! A buffered implementation that automatically groups native prices API
+//! requests into batches.
+
+use {
+    crate::price_estimation::{native::NativePriceEstimateResult, PriceEstimationError},
+    anyhow::anyhow,
+    async_trait::async_trait,
+    futures::{
+        channel::mpsc,
+        future::FutureExt as _,
+        stream::{self, FusedStream, Stream, StreamExt as _},
+    },
+    primitive_types::H160,
+    std::{
+        collections::{HashMap, HashSet},
+        future::Future,
+        num::NonZeroUsize,
+        sync::Arc,
+        time::Duration,
+    },
+    tokio::{sync::broadcast, task::JoinHandle, time::error::Elapsed},
+};
+
+/// Buffered configuration.
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to request.
+    ///
+    /// Specifying `None` means no limit on concurrency.
+    pub max_concurrent_requests: Option<NonZeroUsize>,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+    /// The timeout to wait for the result to be ready
+    pub result_ready_timeout: Duration,
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[allow(dead_code)]
+#[mockall::automock]
+#[async_trait]
+pub trait NativePriceBatchFetcher: Sync + Send {
```

Comment:
nit: This trait could extend the single request version of the trait since it's safe to assume that all APIs can also implement the single request version with a batch size of 1.

Reply:
umm, I don't fully get you. Do you mean to extend the trait to implement the request for one token? I don't think we should do it, because this trait is specifically for bulk requests + the requester can just do `&[token]` for individual request.

Reply:
I mean you could do `pub trait NativePriceBatchFetcher: NativePriceEstimating` since the bulk version is a super set of the single version. Also since these are conceptually the same thing I think it makes sense to have the naming reflect that.

Reply:
Ah, ok, I understand. Oh, sure, definitely!

Reply:
@MartinquaXD actually, there is an issue and it is that `NativePriceEstimating` is not async, but:

```rust
    fn estimate_native_price(
        &self,
        token: H160,
    ) -> futures::future::BoxFuture<'_, NativePriceEstimateResult>;
```
    
is there a reason for it not to be directly `async`? if I can make it `async` it'd be the best, but maybe there is some reason I a missing :thinking:

Reply:
the `async` keyword only says the function returns `impl Future`. I assume it was done like this here because of lifetime annotation limitations of the `async_trait` macro but conceptually this is equivalent.

Reply:
Yup, that's why I was wondering maybe we wanted to keep it in the "not syntactically sugar" version for a reason :thinking: but changing it! :ok_hand: